### ISSUE TITLE
help

### ISSUE THREAD

**Issue by vigo** [tectiv3/lsp-client#1]
hi, how can i integrate this to TextMate? copilot only? can you add more documentation how to install it etc?

**Comment by tectiv3** [comment]
hi! Sorry for the late reply...

well, it is kinda ad-hoc but it works :)

So first of all you need to get vim copilot server and notice where it is installed to put in config.
```
yarn global add copilot-node-server@1.9.3
```
##### config.json
```json
{
    "node_path": "/opt/homebrew/opt/node@20/bin/node",
    "copilot_path": "~/.config/yarn/global/node_modules/copilot-node-server/copilot/dist/agent.js",
    "volar_path": "",
    "gopls_path": "",
    "intelephense_path": "",
    "tsdk_path": "~/.config/yarn/global/node_modules/typescript/lib/",
    "port": "8787",
    "enable_logging": false
}
```

Then build this package which will act as a proxy between copilot server and Textmate, then run it. 
I login into Copilot using VSCodium Github Copilot extension and that seems to work for vim copilot as well.

Install this bundle:
https://github.com/tectiv3/PHP-LSP.tmbundle

Basically you need only those two commands for copilot so you can just copy them from the bundle and add to whatever language bundle you are using:

<img width="800" alt="Screenshot 2024-12-06 at 13 18 38" src="https://github.com/user-attachments/assets/6bcd0560-b4b2-45ea-adab-12af530bccd4" />

<img width="800" alt="Screenshot 2024-12-06 at 13 21 09" src="https://github.com/user-attachments/assets/f504e2db-d84f-4915-9bb7-df6cfb50fe33" />

then hitting option+Esc will generate completion at the cursor.

**Comment by vigo** [comment]
thank you

**Comment by fiedl** [comment]
@vigo Have you managed to get it working?

**Comment by vigo** [comment]
nope not yet. TextMate UI (ruby lib) crashes a lot. I really don't know how to handle. It randomly works and randomly crashes. When it crashes (ui element) TextMate hangs and i have to kill it manually. Development phase getting harder and harder but i'm not finished yet :)

**Comment by tectiv3** [comment]
huh... Works flawlessly for me.
what ruby/textmate version do you have?

I don't remember exactly but I think I went through Textmate's support libraries and updated the hardcoded ruby 1.8 version. 

Maybe thats the reason for your crashes.

**Comment by vigo** [comment]
i'm trying gopls:

<img width="2022" alt="Image" src="https://github.com/user-attachments/assets/d5be7a0d-baa7-431f-871d-648f3eadd73a" />

when i try to complete:

<img width="609" alt="Image" src="https://github.com/user-attachments/assets/4fe99e03-1e12-4a21-a9e2-126f35315a6f" />

**Comment by tectiv3** [comment]
check that your `on open` includes  `source.go`

<img width="689" alt="Image" src="https://github.com/user-attachments/assets/6c2535b0-e574-4944-a1fe-90c5d7218dc7" />

**Comment by vigo** [comment]
```bash
$ curl -X POST http://127.0.0.1:8787/ \
     -H "Content-Type: application/json" \
     -d '{
           "method": "completion",
           "body": {
             "textDocument": { "uri": "file:///Users/vigo/Development/vigo/golang/cli/tablo/main.go" },
             "position": { "line": 14, "character": 16 },
             "context": { "triggerKind": 1 }
           }
         }'
{"message":"not initialized","result":"error"}
``` 

i guess for gopls, it has to be initialized first... i couldn't find how to make requests to gopls. in what order etc.. any idea?

**Comment by vigo** [comment]
@tectiv3 i'm trying to do something here https://github.com/vigo/textmate2-golang if you want to give a hand that would be great

**Comment by tectiv3** [comment]
That is exactly what it is doing in "on open":
<img width="585" alt="Image" src="https://github.com/user-attachments/assets/6dd1dc79-70cf-4b8a-8103-fe8b961312f7" />

**Comment by vigo** [comment]
well, i'm crashed... no idea.. getting "not initialized" error... i don't know what to do

**Comment by tectiv3** [comment]
notice that it checks for `TM_ENABLE_IS` set to true to actually send initalize

**Comment by vigo** [comment]
i commented out those lines... enabled by default

<img width="667" alt="Image" src="https://github.com/user-attachments/assets/b608e95d-a3c2-4881-a5e7-7b062de0cd8a" />

**Comment by tectiv3** [comment]
and you still do not see initialize requests in lsp-client logs?

**Comment by tectiv3** [comment]
look in "completion" command - it can also call initialize every time. you can uncomment that if `on open` events are not firing for you

**Comment by vigo** [comment]
i'm using `go version go1.24.0 darwin/arm64` and no.. nothing

**Comment by vigo** [comment]
i can provide google meet link :) we can check together if you are available? (just trying)

**Comment by tectiv3** [comment]
not completion, sorry, `Show definition` command. 
Look for an example there. Or you can just copy initaliaze syntax from "on open" into `complete`

**Comment by tectiv3** [comment]
sure, if you are still having problems - I am happy to help!

**Comment by vigo** [comment]
hop in https://meet.google.com/uro-gtav-pbr

**Comment by tectiv3** [comment]
<img width="723" alt="Image" src="https://github.com/user-attachments/assets/8c2f9d22-96a7-4a7c-9080-9638a71a703c" />

**Comment by vigo** [comment]
by the way, we nailed it together with @tectiv3 ... stay tuned for great surprises! TextMate 4ever!

**Comment by tectiv3** [comment]
My `gocode` based completion command

```ruby
#!/usr/bin/env ruby

require "open3"

require ENV['TM_SUPPORT_PATH'] + "/lib/ui.rb"
require ENV['TM_SUPPORT_PATH'] + "/lib/escape.rb"
require ENV['TM_SUPPORT_PATH'] + "/lib/exit_codes.rb"
require ENV['TM_SUPPORT_PATH'] + "/lib/web_preview"
require ENV['TM_SUPPORT_PATH'] + "/lib/tm/htmloutput.rb"

# TextMate's special GOPATH used in .tm_properties files prepended to the environment's GOPATH
ENV['GOPATH'] = (ENV.has_key?('TM_GOPATH') ? ENV['TM_GOPATH'] : '') +
                (ENV.has_key?('GOPATH') ? ':' + ENV['GOPATH'] : '').sub(/^:+/,'')

# Call tool to determine gopath
if ENV.has_key?('TM_GO_DYNAMIC_GOPATH')
  Dir.chdir(ENV['TM_DIRECTORY']) do
    ENV['GOPATH'] = `#{ENV['TM_GO_DYNAMIC_GOPATH']}`.chomp
  end
end

gocode = ENV['TM_GOCODE'] || 'gocode'

# load current document from stdin
document = []
while line = $stdin.gets
  document.push(line)
end

# byte offset of cursor position from the beginning of file
cursor = document[ 0, ENV['TM_LINE_NUMBER'].to_i - 1].join().length + ENV['TM_LINE_INDEX'].to_i


# send document to stdin of gocode and read output
output = ""
Open3.popen3("#{gocode} -f=csv autocomplete #{ENV['TM_FILEPATH']} #{cursor}") {|stdin, stdout, stderr|
  stdin.puts document.join
  stdin.close
  output = stdout.readlines.join
}

# quit if no completions found
TextMate.exit_show_tool_tip("No completions found.") if output.length == 0

# set up images for use by DIALOG
# this probably should be done only once, somehow.
icon_plist = "{ " + [ "const", "func", "package", "type", "var" ].map { |v| 
  "#{v} = '#{ENV['TM_BUNDLE_SUPPORT']}/icons/#{v}.pdf';"
}.join(" ") + " }"
system( ENV['DIALOG'], "images", "--register", icon_plist )

# helper function to build the choice hash
def make_completion_hash(line)
  comp = line.split(",,", 3)

  match = comp[1]
  image = comp[0]

  display = " " + comp[1]
  display += comp[0] == "func" ? comp[2].gsub(/^func/, "") : " " + comp[2]
  
  # input : "foo(x func(func())) (z int, k int)"
  # output: "0000111111122222210001111111111110"
  def depth_at_i(sig)
    depths = Array.new
    depth = 0
    sig.chars { |ch|
      depth-=1 if ch == ")"
      depths << depth
      depth+=1 if ch == "("
    }
    return depths
  end

  # returns function arguments in the form "x arg, y arg"
  def get_f_args(sig)
    depths = depth_at_i(sig)
    pos = sig.index(")")
    while pos != nil && depths[pos] > 0
      pos = sig.index(")", pos+1)
    end
    return sig[ Range.new(sig.index("(")+1, pos - 1) ]
  end
  
  def split_args(sig)
    args = Array.new
    depths = depth_at_i(sig)
    start = 0
    pos = sig.index(",")

    while pos != nil
      if depths[pos] == 0
        args << sig[ Range.new(start, pos-1) ]
        start = pos+1
      end
      pos = sig.index(",", pos+1)
    end

    lastarg = sig[ Range.new(start, sig.length) ].strip
    args << lastarg unless lastarg == ""
    return args
  end

  if comp[0] == "func"
    i = 0
    insert = "(" + split_args( get_f_args( display )).map { |m| "${#{i += 1}:"+e_snp(m)+"}" }.join(", ") + ")$0"
    # TextMate.exit_show_tool_tip( { 'match' => match, 'display' =>  display, 'insert' => insert, 'image' => image })
    # exit
  else
    insert = ""
  end
  
  return { 'match' => match, 'display' =>  display, 'insert' => insert, 'image' => image }
end

# build the list of completion choices.
hash = output.split("\n").collect { |v| make_completion_hash( v ) }
# puts hash.inspect
# exit
options = { :extra_chars => "_", :case_insensitive => false }
      # TextMate.exit_show_tool_tip(hash)
# if there is only one match, insert. no need to show the menu
if hash.length == 1
  word = ENV['TM_CURRENT_WORD'] || ""
  snippet = hash[0]["match"].gsub(/^#{Regexp.escape(word)}/, "") + hash[0]["insert"]
  #`"$DIALOG" x-insert --snippet #{e_sh snippet}`
  #full = document.join()
  #print e_sn(full[0..cursor]) + "$0" + e_sn(full[cursor+1..-1])
  TextMate.exit_insert_snippet( snippet )
else
  TextMate::UI.complete( hash , options )
end
```

Turned out I was not using gopls complete command. So previous screenshot is from this command.

Apparently I use go-pls for diagnostics only!

It has been a few years since I set everything up and since it was working I didn't touch it :))

**Comment by tectiv3** [comment]
And since copilot is usually fast and smart enough - I am not hitting completion shortcut either.

And `godef` is useful for "jump to definition" command



```ruby
#!/usr/bin/env ruby

require ENV['TM_SUPPORT_PATH'] + '/lib/textmate'
require ENV['TM_SUPPORT_PATH'] + '/lib/tm/process'
require ENV['TM_SUPPORT_PATH'] + '/lib/tm/save_current_document'
require 'pathname'

# TextMate's special TM_GODEF or expect 'godef' on PATH
godef_cmd = ENV['TM_GODEF'] || 'godef'

TextMate.save_if_untitled('go')

# load current document from stdin
document = []
while line = $stdin.gets
  document.push(line)
end

# byte offset of cursor position from the beginning of file
cursor = document[ 0, ENV['TM_LINE_NUMBER'].to_i - 1].join().length + ENV['TM_LINE_INDEX'].to_i

args = []
args.push(godef_cmd)
args.push('-f')
args.push(ENV['TM_FILEPATH'])
args.push('-o')
args.push("#{cursor}")

# /usr/local/Cellar/go/1.7.1/libexec/src/database/sql/sql.go:960:15
out, err = TextMate::Process.run(*args)

if (err.nil? || err == '') && out
  file_details = out.split(':')
  file = Pathname.new(file_details[0]).realpath
  TextMate.go_to(:file => file,
                 :line => file_details[1].to_i,
                 :column => file_details[2].to_i)
elsif err.casecmp "godef: no identifier found";
  require "#{ENV['TM_SUPPORT_PATH']}/lib/ui.rb"
  package = TextMate::UI.request_string :title => "Open Package", :default => '', :prompt => "Which package do you wish to open?"
  #TODO: Find and open package
else
  TextMate.exit_show_tool_tip(err)
end
```
